<a href="https://colab.research.google.com/github/AilingLiu/Growth_Analysis/blob/master/Product_Recommendation_by_Collaborative_Filtering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings(action="ignore")

In [2]:
url='https://github.com/AilingLiu/Growth_Analysis/blob/master/Data/online_retail.csv?raw=true'
retail = pd.read_csv(url, encoding = 'unicode_escape')
retail['InvoiceDate']=pd.to_datetime(retail['InvoiceDate'])
retail['Payment'] = retail['Quantity'] * retail['UnitPrice']
retail.head()

InvoiceNo StockCode  ...         Country  Payment
0    536365    85123A  ...  United Kingdom    15.30
1    536365     71053  ...  United Kingdom    20.34
2    536365    84406B  ...  United Kingdom    22.00
3    536365    84029G  ...  United Kingdom    20.34
4    536365    84029E  ...  United Kingdom    20.34

[5 rows x 9 columns]

In [4]:
retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 9 columns):
InvoiceNo      541909 non-null object
StockCode      541909 non-null object
Description    540455 non-null object
Quantity       541909 non-null int64
InvoiceDate    541909 non-null datetime64[ns]
UnitPrice      541909 non-null float64
CustomerID     406829 non-null float64
Country        541909 non-null object
Payment        541909 non-null float64
dtypes: datetime64[ns](1), float64(3), int64(1), object(4)
memory usage: 37.2+ MB


In [3]:
retail.describe()

Quantity      UnitPrice     CustomerID        Payment
count  541909.000000  541909.000000  406829.000000  541909.000000
mean        9.552250       4.611114   15287.690570      17.987795
std       218.081158      96.759853    1713.600303     378.810824
min    -80995.000000  -11062.060000   12346.000000 -168469.600000
25%         1.000000       1.250000   13953.000000       3.400000
50%         3.000000       2.080000   15152.000000       9.750000
75%        10.000000       4.130000   16791.000000      17.400000
max     80995.000000   38970.000000   18287.000000  168469.600000

Filter out the negatieve quantity records, and the missing customer ID.

In [5]:
df = retail.loc[(retail['Quantity'] > 0)&(retail['CustomerID'].notnull())]
df.isnull().any()

InvoiceNo      False
StockCode      False
Description    False
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID     False
Country        False
Payment        False
dtype: bool

In [8]:
(df.Quantity<0).any()

False

## User-based Recommendation System

We will first build a user_item matrix, then calculate the user similarities based on the items they bought. If user b has the highest similarity with user a, we will recommend the products that is bought by user b to user a, provided user a has not bought the same products before.

In [11]:
user_item = pd.pivot_table(data=df, 
                               values='Quantity', 
                               index='CustomerID', 
                               columns='StockCode', aggfunc='sum', fill_value=0)
user_item.head()

StockCode   10002  10080  10120  10123C  10124A  ...  C2  DOT  M  PADS  POST
CustomerID                                       ...                        
12346.0         0      0      0       0       0  ...   0    0  0     0     0
12347.0         0      0      0       0       0  ...   0    0  0     0     0
12348.0         0      0      0       0       0  ...   0    0  0     0     9
12349.0         0      0      0       0       0  ...   0    0  0     0     1
12350.0         0      0      0       0       0  ...   0    0  0     0     1

[5 rows x 3665 columns]

In [27]:
from sklearn.metrics.pairwise import cosine_similarity
user_item_similarity = pd.DataFrame(cosine_similarity(user_item))
user_item_similarity.index = user_item.index
user_item_similarity.columns = user_item.index
user_item_similarity.head()

CustomerID  12346.0   12347.0   12348.0  ...   18282.0   18283.0   18287.0
CustomerID                               ...                              
12346.0         1.0  0.000000  0.000000  ...  0.000000  0.000000  0.000000
12347.0         0.0  1.000000  0.148879  ...  0.010873  0.074510  0.108942
12348.0         0.0  0.148879  1.000000  ...  0.000000  0.175170  0.110096
12349.0         0.0  0.020750  0.000169  ...  0.013398  0.065295  0.022576
12350.0         0.0  0.014435  0.000315  ...  0.000000  0.019385  0.000000

[5 rows x 4339 columns]

The diagnal is 1, meaning user compares to itself. We can now take some examples of customer ID and see who they match the most.

In [30]:
user_item_similarity.loc[12346.0].nlargest(5)

CustomerID
12346.0    1.000000
15567.0    0.807065
18242.0    0.384018
17030.0    0.376196
17309.0    0.371391
Name: 12346.0, dtype: float64

Compare user 12346 with 15567 and 18242:

**User 12346**

In [39]:
df.loc[df['CustomerID']==12346.0, ['StockCode', 'Quantity', 'Description']]

StockCode  Quantity                     Description
61619     23166     74215  MEDIUM CERAMIC TOP STORAGE JAR

**User 15567**

In [40]:
df.loc[df['CustomerID']==15567.0, ['StockCode', 'Quantity', 'Description']]

StockCode  Quantity                     Description
485775     23166        96  MEDIUM CERAMIC TOP STORAGE JAR
485776     21401        48              BLUE PUDDING SPOON
485777     21400        48               RED PUDDING SPOON
485778     23178         6                JAM CLOCK MAGNET
485779     22762         1   CUPBOARD 3 DRAWER MA CAMPAGNE
485780     22969        12    HOMEMADE JAM SCENTED CANDLES
485781     23154        12        SET OF 4 JAM JAR MAGNETS

**User 18242**

In [43]:
df.loc[df['CustomerID']==18242.0, ['StockCode', 'Quantity', 'Description']].sort_values(by=['Quantity'], ascending=False)

StockCode  Quantity                     Description
206593     23167        96  SMALL CERAMIC TOP STORAGE JAR 
364406     47566        50                   PARTY BUNTING
206589     22962        48           JAM JAR WITH PINK LID
206594     23166        48  MEDIUM CERAMIC TOP STORAGE JAR
182553     84879        24   ASSORTED COLOUR BIRD ORNAMENT
...          ...       ...                             ...
206600     22283         2        6 EGG HOUSE PAINTED WOOD
206588     21524         2  DOORMAT SPOTTY HOME SWEET HOME
206587     48184         2           DOORMAT ENGLISH ROSE 
206603     22606         1      WOODEN SKITTLES GARDEN SET
206602     22526         1       WHEELBARROW FOR CHILDREN 

[78 rows x 3 columns]

Apparently, user 12346 only bought one item `23166`, but A LOT. The other two users also bought quite some `23166` but not as many as user 12346.

We can compare user `12348.0` to others also and see who he is similar to.

In [46]:
user_item_similarity.loc[12348.0].nlargest(5)

CustomerID
12348.0    1.000000
16174.0    0.494248
14163.0    0.448390
17940.0    0.386058
13623.0    0.372502
Name: 12348.0, dtype: float64

In [56]:
compares = df.loc[df['CustomerID']==12348.0, ['StockCode', 'Quantity', 'Description']].merge(df.loc[df['CustomerID']==16174.0, ['StockCode', 'Quantity', 'Description']], 
                                        how='outer', 
                                        on='StockCode',
                                        suffixes=('_12348', '_16174')
                                        ).fillna(0).sort_values(by='Quantity_12348', ascending=False)
compares.head(15)

StockCode  Quantity_12348  ... Quantity_16174         Description_16174
21     21985           144.0  ...            0.0                         0
23     21983           144.0  ...            0.0                         0
24     21967           144.0  ...            0.0                         0
7      21981           144.0  ...            0.0                         0
8      21982           144.0  ...            0.0                         0
20     21980           144.0  ...            0.0                         0
27     23077           120.0  ...           20.0       DOUGHNUT LIP GLOSS 
3      84991           120.0  ...            0.0                         0
5      21213           120.0  ...            0.0                         0
15     21977           120.0  ...            0.0                         0
13     22952           120.0  ...            0.0                         0
28     23077           120.0  ...          200.0       DOUGHNUT LIP GLOSS 
29     23078            96.0  ...           24.0  ICE CREAM PEN LIP GLOSS 
30     23078            96.0  ...          240.0  ICE CREAM PEN LIP GLOSS 
31     23078            96.0  ...           24.0  ICE CREAM PEN LIP GLOSS 

[15 rows x 5 columns]

With the comparison as above, we can find out the items that user_12348 has not bought but bought by user_16174. We can recommend these products to this user.

## Item-based Recommendation

In [57]:
item_user = user_item.T
item_user_similarity = pd.DataFrame(cosine_similarity(item_user))
item_user_similarity.index = item_user.index
item_user_similarity.columns = item_user.index
item_user_similarity.head()

StockCode     10002  10080     10120    10123C  ...  DOT         M  PADS      POST
StockCode                                       ...                               
10002      1.000000    0.0  0.001548  0.000990  ...  0.0  0.000308   0.0  0.074260
10080      0.000000    1.0  0.000000  0.000000  ...  0.0  0.000006   0.0  0.000000
10120      0.001548    0.0  1.000000  0.004903  ...  0.0  0.007664   0.0  0.000331
10123C     0.000990    0.0  0.004903  1.000000  ...  0.0  0.000000   0.0  0.000000
10124A     0.000000    0.0  0.000000  0.000000  ...  0.0  0.000000   0.0  0.000000

[5 rows x 3665 columns]

Take product 10002 as example, we will check its related product, and recommend them to any user who has bought 10002 before.

In [81]:
related_items = item_user_similarity.loc['10002'].nlargest(10)
related_prods = df.loc[df['StockCode'].isin(related_items.index), ['StockCode', 'Description']].drop_duplicates().set_index('StockCode').join(related_items)
related_prods.rename(columns={'10002': 'cosSim'}, inplace=True)
related_prods.sort_values(by='cosSim',ascending=False)

Description    cosSim
StockCode                                            
10002           INFLATABLE POLITICAL GLOBE   1.000000
10125               MINI FUNKY DESIGN TAPES  0.853890
23224          CHERUB HEART DECORATION GOLD  0.713423
23222           CHRISTMAS TREE HANGING GOLD  0.699006
85014A         BLACK/BLUE POLKADOT UMBRELLA  0.626077
20682      RED RETROSPOT CHILDRENS UMBRELLA  0.597314
23007                SPACEBOY BABY GIFT SET  0.575727
23009           I LOVE LONDON BABY GIFT SET  0.557707
23010           CIRCUS PARADE BABY GIFT SET  0.553042
85014B               RED RETROSPOT UMBRELLA  0.457680

Customers bought INFLATABLE POLITICAL GLOBE are also likely to buy MINI FUNKY DESIGN TAPES, and CHERUB HEART DECORATION GOLD, which are all some party decoration stuffs. Great, we can also recommend these to custmers.